rfcs/0017-relationships-from-nested-fields.md

```diff
@@ -0,0 +1,117 @@
+        }
+    },
+    "collection_relationships": {
+        "album_publisher": {
+            "arguments": {},
+            "column_mapping": {
+                "publisher_id": "id"
+            },
+            "relationship_type": "object",
+            "target_collection": "publishers"
+        }
+    }
+}
+```
+
+This sort of query is not documented, and needs some clarification.
+
+Note in particular that the relationship column mapping is evaluated in the context of the nested object on the source side: we refer to `publisher_id`, which is a column on the nested `album` object, not on the "current row", which will be at the top of the scope stack. 
+
+This feature therefore ties in with the work on named scopes: we might want to consider keeping the ancestor scopes around on the stack as we descend into nested objects.
+
+## Proposal
+
+The proposal is to document the following clearly in the specification: connectors should _replace the element at the top of the scope stack_ (that is, the "current row") with the current nested object under consideration, each time evaluation descends into a `NestedObject`.
+
+Specifically, column mappings will be evaluated in the context of the nested object's fields, and any other behaviors which rely on the current row will change accordingly:
+
+- `RelationshipArgument::Column` will resolve to a field of the current nested object.
+
+A note needs to be added to the specification to clarify that relationship mappings and arguments should read off the top of the scope stack.
+
+Advantages are:
+
+- This is supported in NDC today.
+- This design allows us to express relationships from objects inside (possibly deeply-nested) nested arrays and objects.
+- The related object appears in the response in the correct place, next to the related source object.
+ 
+The main disadvantage is that we can only express join conditions which make sense in the context of the nested object. Named scopes will go some way towards helping here, but we won’t be able to express join conditions which use column mappings unless the columns on the source side are nested under the current object.
+
+## Alternative Designs
+
+- Instead, we can move the related objects to the top of the query, and change the column mapping data structure to allow mappings between nested fields (using field paths like we have elsewhere).
+  - We would be able to express more join conditions, but we would lose the ability to express relationships from data inside nested arrays. Related data would also not appear next to its counterpart on the source side.
+
+## Future Work
+
+1. Extend column mappings to allow references to nested object fields.
+1. Extend column mappings to allow references to [named scopes](./0015-named-scopes.md).
+
+## Open Questions
+
+- Should we replace the top of the scope stack, or push each nested object in turn?
+  - Pushing means we now have two types of "stack frame" on the scope stack. It might get confusing.
```

Comment:
I just mean that we have two contexts where we push a row onto the stack: exists, and now nested objects. The meaning is probably going to be clear because they're both just rows, but the indexing might get confusing unless we document it very clearly.

Reply:
Perhaps we could label the scopes separately to make it clear? Instead of `{1}`, it could be `{n1}` for nested object scope and `{p1}` for predicate scope?

Reply:
@daniel-chambers I've added my current thinking in a notes section, but I think basically this is going to be unimportant since the scope stack isn't going to be interesting in these cases anyway. It's more about how we define the concept of "current row".